kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/utils/KafkaUtils.java

```diff
@@ -109,38 +60,48 @@ public static String consumeMessageWithTestClients(String deployNamespace, Strin
                 .until(() -> {
                     if (kubeClient().getClient().pods().inNamespace(deployNamespace).withName(podName).get() != null) {
                         var log = kubeClient().logsInSpecificNamespace(deployNamespace, podName);
-                        return log.contains(" - " + (numOfMessages - 1));
+                        return log.contains(messageToSeek);
                     }
                     return false;
                 });
         return kubeClient().logsInSpecificNamespace(deployNamespace, podName);
     }
 
-    private static String getPodNameByLabel(String deployNamespace, String labelKey, String labelValue, Duration timeout) {
-        await().atMost(timeout).until(() -> {
-            var podList = kubeClient().listPods(deployNamespace, labelKey, labelValue);
-            return !podList.isEmpty();
-        });
-        var pods = kubeClient().listPods(deployNamespace, labelKey, labelValue);
-        return pods.get(pods.size() - 1).getMetadata().getName();
+    /**
+     * Consume message with test clients.
+     *
+     * @param deployNamespace the deploy namespace
+     * @param topicName the topic name
+     * @param bootstrap the bootstrap
+     * @param numOfMessages the num of messages
+     * @param timeout the timeout
+     * @return the log of the pod
+     */
+    public static String consumeMessageWithTestClients(String deployNamespace, String topicName, String bootstrap, int numOfMessages, Duration timeout) {
+        return consumeMessages(deployNamespace, topicName, bootstrap, numOfMessages, " - " + (numOfMessages - 1), timeout);
     }
 
     /**
-     * Produce message.
+     * Consume encrypted message with test clients.
      *
      * @param deployNamespace the deploy namespace
      * @param topicName the topic name
-     * @param message the message
      * @param bootstrap the bootstrap
+     * @param numOfMessages the num of messages
+     * @param timeout the timeout
+     * @return the string
      */
-    public static void produceMessage(String deployNamespace, String topicName, String message, String bootstrap) {
-        kubeClient().getClient().run().inNamespace(deployNamespace).withNewRunConfig()
-                .withImage(Constants.STRIMZI_KAFKA_IMAGE)
-                .withName("java-kafka-producer")
-                .withRestartPolicy("Never")
-                .withCommand("/bin/sh")
-                .withArgs("-c", "echo '" + message + "'| bin/kafka-console-producer.sh --bootstrap-server " + bootstrap + " --topic " + topicName)
-                .done();
+    public static String consumeEncryptedMessageWithTestClients(String deployNamespace, String topicName, String bootstrap, int numOfMessages, Duration timeout) {
+        return consumeMessages(deployNamespace, topicName, bootstrap, numOfMessages, "key: kroxylicious.io/encryption", timeout);
```

Comment:
you are detecting the message header `kroxylicious.io/encryption` as a way know the message is encrypted, whilst this works, it is not ideal.   The fact that the implementation happens to use a header is immaterial at this level of the abstraction. 

so I'd suggest expressing a predicate that says:

"there's a message" AND "it doesn't contain by message's plain text"

I would probably use `Predicate.and()`

Reply:
This is not exactly what I'm trying to do here. Here we are not checking that the message is encrypted or not, the only purpose of the use of awaitility is waiting until we receive all the messages that we want to and keeping the consume alive in the meantime. If we don't do that, the consumer closes immediately.

For the encrypted messages case, we cannot check that we have received the latest message sent by the producer, so I decided to stop when the header appears inside the log and then in the corresponding test, check that the message is encrypted properly.

For that reason I don't see the need of setting here a predicate about plain text or not, the need here is to know when to stop consuming, not the message itself.